### ISSUE TITLE
Port TCPNJE support from spinhawk to SDL Hyperion

### ISSUE THREAD

**Issue by ivan-w** [SDL-Hercules-390/hyperion#280]
Port Peter Coghlan's spinhawk TCPNJE 1.0 version from spinhawk to the current SDL version

**Comment by PeterCoghlan** [comment]
This is coming as a surprise to me.  Last month, I was copied on an email which suggested I was in discussions about this happening which was also news to me.

My original intention was to provide a common codebase for tcpnje which would be compatible with and usable with all Hercules variants and which would work on the widest possible range of host platforms without changes and if I had been successful, no porting would have been required, it would just work.

Unfortunately, I have been unable to keep up with incompatibilities and changes detrimental to portability which have been introduced in SDL Hyperion and therefore having common code which works across all Hercules variants has become an impossibility for me.

I have been unable to build SDL Hyperion since sometime around 2018. I am not in a position to provide support for tcpnje in SDL Hyperion. I would like all indications that I might provide support for this arrangement removed from any code and documentation included with SDL Hyperion.

**Comment by Fish-Git** [comment]
@PeterCoghlan

Peter,

I cannot remember exactly when it was nor exactly _what_ it was, but I do specifically remember you mentioning an incompatibility between SDL Hyperion and the other Herculeses (*) out there at some point in the recent past (last year or two?), which I also specifically remember having _fixed_ for you!

What incompatibilities are you now talking about? If there are incompatibilities in SDL Hyperion that is preventing your tcpnje driver from building or working correctly with SDL Hyperion, you should create a GitHub Issue for it/them so that we (myself and the rest of the SDL Hyperion team) can work on trying to get them fixed for you!

Thanks.

_____
(*) _How the heck do you pluralize "Hercules"?_

**Comment by Fish-Git** [comment]
> I would like all indications that I might provide support for this arrangement removed from any code and documentation included with SDL Hyperion.

I am not aware that there is any.  But if you can find any I'd be happy to remove it.  Just point it out and it'll be gone.

**Comment by PeterCoghlan** [comment]
Fish,

You fixed what feels like dozens of incompatibilities and portability issues for me dating back to well before SDL Hyperion came to exist.  The problem for me is these issues continue to arise quicker than I can spot them, report them and ask for them to be fixed.  I don't want to be on that treadmill.

The biggest issue for me is that support for static linked executables was removed from SDL Hyperion source (in 2018 I think). This made it practically impossible to build SDL Hyperion on any platform which is neither Windows nor one of a limited range of Unix variants.  Since then, as far as I can see, it has only been possible to use one of two highly platform specific methods of generating dynamically linked executables, a Windows specific method and a very non-portable Unix specific method coded in ltdl.c / hdl.c. Neither of these methods of building dynamically linked executables can be used in my environment without major surgery which would be all the more difficult due to the almost complete lack of descriptive comments in ltdl.c in particular.  (I tried and gave up.)  It also makes it so much more difficult to share any common code between SDL Hyperion and any other Hercules variants.

After my experience last year asking for a tiny incompatibility to be fixed (which you did fix - thanks) I felt it would be a mammoth task to ask for the restoration of support for statically linked executables [1] and even if this happened, being unable to build SDL Hyperion for so long now, I would not be aware of what new difficulties may have arisen for me since the last time I could build it, which from previous experience could be quite a few.

[1] I only made use of the support for statically linked executables present in the actual C source code which I think amounted to a few hundred lines of preprocessor macros and suchlike across all the codebase.  (It was probably hard work to remove them and is likely nigh on impossible to replace them all correctly now.) I never used any of the supplied build procedures, makefiles, configure scripts, shell scripts, cmd/bat files, libtools or anything else like that which are all rumoured to not be able to properly generate statically linked executables anyway.

**Comment by PeterCoghlan** [comment]
> >  I would like all indications that I might provide support for this arrangement removed from any code and documentation included with SDL Hyperion.
> 
>  I am not aware that there is any. But if you can find any I'd be happy to remove it. Just point it out and it'll be gone.
> 

There are (probably) not any statements involving me in the current SDL Hyperion codebase that I would object to but there will be if the proposal in the title of this thread goes ahead.

**Comment by ivan-w** [comment]
One of the thing I ran into is that there is a mixture of the use of "TCPNJE" as if it where a "typedef" and "struct TCPNJE". Possibly some compilers self correct this, but gcc doesn't. Once I fixed that, there was only a couple of compilation errors that were easy to fix (mostly the missing definition of the variable "filename" which is implicitly used in a convenience macro.

So now it compiles, but I haven't tried using it yet.

**Comment by ivan-w** [comment]
What the.... a couple tweaks here and there and.. at least I have the "TCPNJE" device Online to RSCS... it's not connecting yet, but that must be something else. But all the module loading and such is resolved (at least under linux).

**Comment by ivan-w** [comment]
Auxiliary question : Would the VM/370 NJE modification support CTC instead of BSC ?

**Comment by ivan-w** [comment]
So far I am getting from my TCPNJE capable RSCS subsystem this message when connecting to the TCP/NJE driver : 
04:02:07 DMTTNE945E Protocol error detected by link VMF, code 1818 -- link being deactivated 
04:02:07 DMTTNE945E Link deactivated -- data being received exceeds expected length          
04:02:07 DMTTNE183I Link VMF session terminated                                              
04:02:07 DMTMAN002I Link VMF deactivated

**Comment by PeterCoghlan** [comment]
> One of the thing I ran into is that there is a mixture of the use of "TCPNJE" as if it where a "typedef" and "struct TCPNJE". Possibly some compilers self correct this, but gcc doesn't. 

I don't see why this should be an issue.  The same is true for SYSBLK, REGS and various other typedefs / structs or at least it was true the last time I was able to build this variant.

> Once I fixed that, there was only a couple of compilation errors that were easy to fix (mostly the missing definition of the variable "filename" which is implicitly used in a convenience macro.

I expect this is a result of some incompatible code divergence in SDL Hyperion.

> Auxiliary question : Would the VM/370 NJE modification support CTC instead of BSC ?

No.

**Comment by ivan-w** [comment]
Nowhere is "TCPNJE" typedef'd... but there are plenty of portion of the code that assume it is.

**Comment by PeterCoghlan** [comment]
> Nowhere is "TCPNJE" typedef'd..

It is in htypes.h

**Comment by ivan-w** [comment]
It's a shame the RSCS NJE mod doesn't support CTC... it'd be closer to the spirit when doing TCP/NJE

**Comment by ivan-w** [comment]
Not in the current version.. htypes.h does NOT define TCPNJE as a typedef !

**Comment by ivan-w** [comment]
htypes.h contains stuff that are pertinent to multiple components of hercules or as a whole... "struct TCPNJE" is a control block that is only relevant to the tcp/nje handler... it doesn't (and shouldn't) contain any information that is relevant to any other portions of the engine.

**Comment by PeterCoghlan** [comment]
> It's a shame the RSCS NJE mod doesn't support CTC... it'd be closer to the spirit when doing TCP/NJE

This was always my preferred option.  I have an RSCS mod to support CTCA close to working properly but the protocol standard RSCS uses across the CTCA is not sufficient to keep things synchronised between RSCS and a TCPNJE device.  BSC lines support ENABLE and DISABLE CCWs which allow additional information to be conveyed between the two.

**Comment by PeterCoghlan** [comment]
> htypes.h contains stuff that are pertinent to multiple components of hercules or as a whole... "struct TCPNJE" is a control block that is only relevant to the tcp/nje handler... it doesn't (and shouldn't) contain any information that is relevant to any other portions of the engine.

This could well be true and if so probably should be fixed.  However, TCPNJE is based on COMMADPT and there is a typedef COMMADPT in the htypes.h I am looking at, not far above where typedef TCPNJE is located.

**Comment by ivan-w** [comment]
The fact that COMMADPT is in htypes.h is a bug I need to fix ! It shouldn't be there ! I probably made this error 15+ years ago :P

**Comment by PeterCoghlan** [comment]
> The fact that COMMADPT is in htypes.h is a bug I need to fix ! It shouldn't be there ! I probably made this error 15+ years ago :P I was starting - made errors... (can't deny it)...

Ivan,

You are being too hard on yourself.  I think if you remove typedef COMMADPT from htypes.h you will probably find that anything that includes hstructs.h but does not include commadpt.h (which is nearly everything) will fail to compile.

To deal with this issue properly, struct DEVBLK should be redesigned to allow device type independent information and device type specific information to be properly separated.  However, doing that would create all sorts of incompaibilities with everything else out there so this is another of these situations when it would be better to just let sleeping dogs lie rather than try to fix something that isn't really broken.

**Comment by ivan-w** [comment]
Added a slightly modified version of the TCP/NJE support based on Pete's version. I just changed all reference to the TCPNJE typedef to "struct TCPNJE" and everything compiles just fine.

tcpnje.c and tcpnje.h (and the necessary changes to Makefile.am) where added to the current version (also added a .md version of the README.TCPNJE)

**Comment by ivan-w** [comment]
Fish, would you be kind enough to integrate it with the Doze build ?

**Comment by PeterCoghlan** [comment]
Ivan, I think the least you could do is to honour the request I made in my first response to this issue, that you remove any indications that I might provide support for this arrangement from the code and documentation.  I particularly would like the email address I included removed.

**Comment by Fish-Git** [comment]
> Fish, would you be kind enough to integrate it with the Doze build?

10-4.

> If any problem arises please ping me!

10-4!

**Comment by Fish-Git** [comment]
FYI to Ivan (and others):

Main README document:

* https://github.com/SDL-Hercules-390/hyperion/blob/master/README.md

(which can be reached by scrolling down to the bottom of our main repository web page at https://github.com/SDL-Hercules-390/hyperion).

Notice that there's a "[**Adding New Files to Hercules**](https://github.com/SDL-Hercules-390/hyperion/blob/master/readme/README.ADDING.md)" link in the "**HERCULES INTERNAL READMEs**" section which documents what needs to be done:

* https://github.com/SDL-Hercules-390/hyperion/blob/master/readme/README.ADDING.md

You did the *Nix side just fine! You just did't do the Windows (MSVC) stuff, that's all. But I'll take care of it for you this time.

The changes needed for Windows basically involve updating the MSVC makefile fragments (e.g. OBJ_CODE.msvc, MODULES.msvc and MOD_RULES2.msvc) and the various Visual Studio files (e.g. Hercules_VSnnnn.vcxproj, Hercules_VSnnnn.vcxproj.filters, etc). That's about it.

But don't worry about it. I'll take care of that for you this time. I just wanted you (and others) to be aware that there _is_ a documented procedure that explains what needs to be done for the Windows side of things. I created so that you guys can do it yourself without me having to always do it, just in case anything should happen to me.

But as I said, don't worry about it this time. I'll take care of it. I just wanted you to be aware of the documented procedure...

_...for next time!_ &nbsp; `;-)`

I'll go ahead and add a new entry for your new TCPNJE readme to the "Features and Operation" section of the main README too, which it looks like you also forgot to do. But it's no big deal. I'll do it.

**Comment by Fish-Git** [comment]
> Ivan, I think the least you could do is to honour the request I made in my first response to this issue, that you remove any indications that I might provide support for this arrangement from the code and documentation. I particularly would like the email address I included removed. It would be even better, if you could remove my name altogether.

I'll take care of it. At least in the README anyway. I haven't looked at actual source code files yet. Is there anything there I need to get rid of?

**Comment by ivan-w** [comment]
I don't get all the crazy stuff.. turning typedef'd TCPNJE names to their original "struct TCPNJE" name is purely cosmetic. The two macro defintions at the begining are just there because this version is only intended for inclusion in the SDL version. There was absolutely no other change in the core of the logic (which is yours)... But hey.. I don't really care - I just want to get something that works ! If you want to have a unified version that works on any and all forks that exist, fine by me !

**Comment by PeterCoghlan** [comment]
Thanks Fish.

Near the end of the banner comment at the top of tcpnje.c

**Comment by ivan-w** [comment]
I'm not even sure my .md file is github compliant !

**Comment by PeterCoghlan** [comment]
Ivan,

This has nothing to do with typedefs.  I disagree fundamentally with the direction SDL Hyperion has taken on questions of compatibility, portability and documentation of significant changes.  I do not want to be appearing to endorsing this direction, I do not want to contribute to this but it appears I have no option on that and I certainly do not want to want to be providing support for it.

**Comment by ivan-w** [comment]
Peter,

 Fine by me - but I find the direction correct and much more flexible. But that's a question that is irrelevant to the code at hand. the TCP/NJE support is still under an Open Source Initiative license I believe. I see no reason not to keep credit for the work you performed on that particular element. But it is your choice that any reference to you be removed from the current version.

**Comment by Fish-Git** [comment]
> I'm not even sure my .md file is github compliant!

It mostly was. But I fixed it up for you. &nbsp; `:)`

_(I haven't committed anything yet but am about to...)_

**Comment by ivan-w** [comment]
Thank ou my good friend !

**Comment by ivan-w** [comment]
I really need to install VScode again on my machine (with the .md linter)

**Comment by Fish-Git** [comment]
> I'm not even sure my .md file is github compliant!

FYI:

* https://guides.github.com/features/mastering-markdown/
* https://chrome.google.com/webstore/detail/markdown-viewer/ckkdlimhmcjmikdlpkmbgfkaikojcbjk

I have the PDF saved from the first link. I don't refer to it as often as I used to, but I still do on occasion (like when I need to turn something into a table).

The second link is to a Chrome browser extension that I use to allow me to visually "see" what a `.md` markdown file is going to actually look like in a browser. Makes it a _LOT_ easier to make sure your markdown is right.

Don't know if there's one for FireFox or not but I would be surprised if there wasn't.

**Comment by ivan-w** [comment]
I usually use VScode for this now (yes, I am a linux guy - but for my day to day work I use a Windows workstation)

**Comment by Fish-Git** [comment]
> I usually use VScode for this now

I've heard of that, yes, but never bothered to look into it. I have VS2008 _(still!!)_ and it works fine for me. I have VS2017 installed in a VMware virtual machine but I don't really use it except for quick tests to make sure Herc still builds okay with it.

> MS has gone a long way!

Yep! ... Though I _still_ hate the new VStudio's user interface. Almost as much as I hate the Windows 8 & 10 "Metro" user interface.

I used to make func (?) of the people that refused to use Windows 7, insisting on sticking with Windows XP instead, but ever since Windows 8 and then 10 came out I changed my tune. I'm sticking with Windows 7 forever! (*)

___
(*) I haven't purchased it yet but I'm going to pay for extended support and when that runs out, use "[**0patch**](https://0patch.com/)".

**Comment by ivan-w** [comment]
VScode is like.. Eclipse+++++++++

**Comment by Fish-Git** [comment]
> Near the end of the banner comment at the top of tcpnje.c

Haven't committed it yet, but the new banner (last few lines) now looks like this:

```C
/* Original Author of commadpt: Ivan Warren                          */
/* TCPNJE hacked from commadpt by Peter Coghlan and integrated into  */
/* SDL Hercules Hyperion by Ivan Warren.                             */
/*                                                                   */
/* PLEASE DO NOT CONTACT PETER COGHLAN FOR SUPPORT! This is a SDL    */
/* Hercules Hyperion implementation that is NOT supporter by Peter!  */
/* If you need support or wish to report a bug, please do so via     */
/* the official SDL Hercules Hyperions GitHub Issues web page at:    */
/*                                                                   */
/*    https://github.com/SDL-Hercules-390/hyperion/issues            */
/*                                                                   */
/* For more information regarding this implementation please refer   */
/* to our README.TCPNJE document.
/*-------------------------------------------------------------------*/
/* TCPNJE version 1.00                                               */
/*-------------------------------------------------------------------*/
```

I trust that meets with your approval?

**Comment by ivan-w** [comment]
Refer to readme/README.TCPNJE.md ;)

**Comment by PeterCoghlan** [comment]
Thanks Fish.  There is a small typo - supporter vs supported but other than that it is fine.

**Comment by ivan-w** [comment]
Peter,
I am still a supporter of your work !

**Comment by Fish-Git** [comment]
> _(I haven't committed anything yet but am about to...)_

Committed!

_(FYI: committed  to **hyperion-txf** too! (*)(**))_
___
(*) It's important to keep our **hyperion-txf** development repository "in sync" with our main hyperion repository, so that when we eventually release TXF (once it's done and fully tested) by renaming the repositories (hyperion ==> hyperion-old, hyperion-txf ==> hyperion), the end result should be the only change that is introduced are the new TXF related changes, and nothing else. That is to say, we don't want to accidentally _lose_ any changes or fixes we made to our main hyperion repository during our TXF development effort. We don't want to "regress" anything when TXF is eventually released.

(**) Keeping hyperion-txf in sync with hyperion can _sometimes_ be a PITA depending on the change, but if you're careful and make sure all of the "changes" you make (commit) to one you also make (commit) to the other, we should be safe. You might have make your changes/commits to txf  manually from time to time (depending on what file was changed and whether it conflicts with any TXF stuff), but such manually synchronization, so far, hasn't been that tough to do.

But that's yet another reason why I do't want our TXF effort to drag on for too long!

**Comment by Fish-Git** [comment]
_**STILL TO DO:**_

Fix the `DBGMSG` macro and all of the "logmsg"s to `WRMSG` instead!

_(Yeah, it's gonna be a pain! But it's gotta be done too!)_

**Comment by PeterCoghlan** [comment]
> Fine by me - but I find the direction correct and much more flexible.

We will have to agree to differ on that then.

> But that's a question that is irrelevant to the code at hand. the TCP/NJE support is still under an Open Source Initiative license I believe.

Perhaps but if it becomes part of an SDL Hyperion release, which seems to be the direction things are going, I believe that will be in contravention of the license under which I contributed my work to the Hercules project.

I find it deeply ironic that you are so enthusiastic to make use of my contribution yet you do not want to foster an environment that makes this kind of contribution possible.  If SDL Hyperion was the only development environment available to me, I simply would not have been able to write the code I wrote.

> I am still a supporter of your work!

How?

**Comment by Fish-Git** [comment]
> _**STILL TO DO:**_
> 
> Fix the `DBGMSG` macro and all of the "logmsg"s to `WRMSG` instead!

Ivan, can you test the attached patch for me? Thanks!

* [280.patch.txt](https://github.com/SDL-Hercules-390/hyperion/files/4272270/280.patch.txt)

**Comment by Fish-Git** [comment]
> ... if it becomes part of an SDL Hyperion release [...] I believe that will be in contravention of the license under which I contributed my work to the Hercules project.

Huh? How so? Please explain!

SDL Hyperion is bound by the same Hercules license that every other Hercules is, or has been, bound to, past, present or future!

> If SDL Hyperion was the only development environment available to me, I simply would not have been able to write the code I wrote.

Why the heck not?! &nbsp; `:(`

**Comment by PeterCoghlan** [comment]
> > ... if it becomes part of an SDL Hyperion release [...] I believe that will be in contravention of the license under which I contributed my work to the Hercules project.
> 
> Huh? How so? Please explain!
> 
> SDL Hyperion is bound by the same Hercules license that every other Hercules is, or has been, bound to, past, present or future!
> 

Well, in the license in question referenced here:

https://github.com/rbowler/spinhawk/blob/master/COPYRIGHT

and here:

https://github.com/SDL-Hercules-390/hyperion/blob/master/COPYRIGHT

(which both seem identical to me), it says (amongst other things):

You may copy and distribute the Software in unmodified form. You may make modifications to the Software and distribute your modifications, in a form that is separate from the Software, such as patches. You may distribute machine-executable forms of the Software or machine-executable forms of modified versions of the Software.

It does not say you may make modifications to the Software and distribute the modified version of the Software.

> > If SDL Hyperion was the only development environment available to me, I simply would not have been able to write the code I wrote.
> 
> 
> Why the heck not?!   :(

For one thing, I can't build SDL Hyperion on the hardware and software platforms which I can build Hercules 3.11/3.12/3.13 on and used to develop TCPNJE on.

**Comment by Fish-Git** [comment]
> It does not say you may make modifications to the Software and distribute the modified version of the Software.

I am not a lawyer, but I believe it does:

> _"You may make modifications to the Software and distribute your modifications..._

And as far as _"separate from"_ and _"patches"_ goes:

> _...in a form that is separate from the Software, such as patches."_

I believe using a source code repository that tracks every change made to any module legally qualifies as distributing modifications to the software in the form of patches.

So I honestly cannot see how we are violating the license in any way. _&nbsp;&lt;shrug&gt;_

But again, I am not a lawyer.

**Comment by PeterCoghlan** [comment]
> I believe using a source code repository that tracks every change made to any module legally qualifies as distributing modifications to the software in the form of patches.
> 
> So I honestly cannot see how we are violating the license in any way.  <shrug>
> 
> But again, I am not a lawyer.

I am not a lawyer either but my impression is the same, that having the Software or a modified version of the Software sitting in a source code repository does not violate the license.  

Where I forsee a difficulty complying with the license is when a modified version of the Software is packaged into _a release_ and distributed in that manner.  As far as I can see, this specific activity constitutes distributing a modified version of the Software where the modifications are not "in a form that is separate from the Software, such as patches" and this activity is therefore not covered by the license.

**Comment by Fish-Git** [comment]
> Where I foresee a difficulty complying with the license is when a modified version of the Software is packaged into _a release_ and distributed in that manner.

How so? Item 4 of the license clearly states that _"machine-executable forms of modified versions of the Software"_ may be distributed (which is what I presume you mean by "release", i.e. "release" = "machine-executable form").

> As far as I can see, this specific activity constitutes distributing a modified version of the Software where the modifications are not "in a form that is separate from the Software, such as patches" ...

Quite true.

And your point regarding this is . . . . . . . . . . . . . . . . . . . what?

> ... and this activity is therefore not covered by the license.

Sure it is! Item 4!

I see absolutely no license violation whatsoever. None. Nada.

**Comment by Fish-Git** [comment]
Ivan (@ivan-w),

Fish wrote:
> Fish wrote:
> > _**STILL TO DO:**_
> > Fix the `DBGMSG` macro and all of the "logmsg"s to `WRMSG` instead!
> 
> Ivan, can you test the attached patch for me? Thanks!
> 
> * [280.patch.txt](https://github.com/SDL-Hercules-390/hyperion/files/4272270/280.patch.txt)

Have you had a chance to test my patch yet? It seems to work fine for me, but I wanted you to give it your approval before committing it. Or should I go ahead and commit it without your testing it? Or is there another way you want to do it? (such as _not_ doing it at all?) Let me know! Thanks.

**Comment by ivan-w** [comment]
DNBFS : 
/home/ivan/hercules/src/fish-hyperion-tcpnje/tcpnje.c:151:63: error: expected expression before ')' token
 #define logmsg( _fmt, ... )     tcpnje_msg( _fmt, __VA_ARGS__ )

**Comment by Fish-Git** [comment]
> DNBFS:

_**DNBFS?**_ &nbsp; What the heck does _that_ mean?

(Sheesh! gcc/clang is such a royal PITA! It compiles fine on Windows!)

**Comment by Fish-Git** [comment]
(FYI: I was specifically _trying_ to prevent modifying _any_ of Peter's code! But thanks to gcc/clan it looks like we might have no choice!)

**Comment by ivan-w** [comment]
DNBFS stands for : Does Not Build From Source

**Comment by PeterCoghlan** [comment]
> How so? Item 4 of the license clearly states that "machine-executable forms of modified versions of the Software" may be distributed (which is what I presume you mean by "release", i.e. "release" = "machine-executable form").

I don't see a complete correspondance between a release and a machine-executable form.  I would argue that a machine-executable form would likely not include source and documentation whille a release may include these things.

Anyway, whether a release involves a machine-executable form or not, I believe the key thing is that the license requires that source is provided in all cases.  My point is that if the source is provided in some sort of archive format in the release, a zip file for example which contains just the source of the modified form of the Software, this does not meet the requirements of the license in item 3 of the granted rights which requires the Software and modifications to it to be separate.

In the specific case of machine-executable forms of modified versions of the Software, restriction c to granted rights item 4 is very clear where it says "You must ensure that all modifications included in the machine-executable forms are available under the terms of this license."

**Comment by Fish-Git** [comment]
> ... I believe the key thing is that the license requires that source is provided in all cases. [...]

Source, both original and modified, is available to anyone at any time for free via our git source code repository.

> In the specific case of machine-executable forms of modified versions of the Software, restriction c to granted rights item 4 is very clear where it says _"You must ensure that all modifications included in the machine-executable forms are available under the terms of this license."_

And they are. What leads you to believe otherwise?

> My point is that if the source is provided in some sort of archive format in the release, a zip file for example which contains just the source of the modified form of the Software, this does not meet the requirements of the license in item 3 of the granted rights which requires the Software and modifications to it to be separate.

They are separate.

I honesty don't understand your complaint at all, Peter. I'm very sorry, but you're making no sense to me whatsoever.

You appear to me to be trying desperately to grasp at what are essentially non-existent straws. &nbsp;`:(`

I mean, I'd like very much to make you happy, but I'm having trouble understanding what it is that you're unhappy about or what you feel we should do to make you happy. &nbsp;`:(`

Does anyone else know?

**Comment by PeterCoghlan** [comment]
> Source, both original and modified, is available to anyone at any time for free via our git source code repository.

I can readily see the modified source in the source code repository as it is right now.  How do I get to see the source that corresponds to a particular release (not that there is any requirement for this to be available)?  More pertiently, how do I get to see the original unmodified source and how do I get to see the modifications made to it for a particular release? Perhaps this is obvious to more seasoned github users and I am insufficiently well versed in the details of how to use this medium, in which case please forgive my ignorance and point me to the documentation where it says how to access these things.

(Aside from any license requirement, I would greatly appreciate being able to view a searchable change log which includes a meaningful description of what changes were made and when.  It would make it so much easier to track down when things went awry.  As far as I know, there used to be one but it was taken away in favour of a link to the git repository which I find to be a complete nightmare to search for particular modifications, unless of course I am failing to understand how to do it properly?)

> > In the specific case of machine-executable forms of modified versions of the Software, restriction c to granted rights item 4 is very clear where it says "You must ensure that all modifications included in the machine-executable forms are available under the terms of this license."
> 
> And they are. What leads you to believe otherwise?
> 

That the terms of the license state that the modifications must be separate from the Software and I can't see how they are separate. (For clarity, I am of course referring to the modifications to the source here, not any modifications to the machine-executable form which neither me nor I believe the license expects anyone to be able to see directly.) Maybe the answer is right in front of me and I just don't know how to work this github thing well enough?


> > My point is that if the source is provided in some sort of archive format in the release, a zip file for example which contains just the source of the modified form of the Software, this does not meet the requirements of the license in item 3 of the granted rights which requires the Software and modifications to it to be separate.
> 
> They are separate.
> 

Forgive me if I am failing to understand something obvious here but I don't follow how a zip file containing modified source code and no patches (or other forms of descriptions of modifications that I can see) can be said to contain separate Software and modifications?

> You appear to me to be trying desperately to grasp at what are essentially non-existent straws.  :(

I didn't want to get sidetracked into this license issue but Ivan brought it up and I responded with my thoughts on the matter.   Then you weighed in and disagreed with me which is fair enough but this has allowed Ivan to duck out of responding to my further comments in the same posting (the one ending in "How?").

> I mean, I'd like very much to make you happy, but I'm having trouble understanding what it is that you're unhappy about or what you feel we should do to make you happy.  :(

One thing that is making me unhappy is that Ivan has dropped out of the conversation I was having with him.  Ivan, can you please respond to the points I made in my posting much further back that ends in "How?"?  While you are at it, could you also say something in response to what I said in my very first posting in this thread about the email I received about two months ago stating I was in some sort of discussions with you and Fish regarding this topic, discussions which I don't recall actually happening?

**Comment by Fish-Git** [comment]
Ivan (@ivan-w),

> DNBFS stands for: Does Not Build From Source

Ah. I never heard that before.

Regardless, please try the below slightly revised version of my patch instead:

* [280.patch.txt](https://github.com/SDL-Hercules-390/hyperion/files/4283350/280.patch.txt)

It now builds (and runs) just fine for me on my CentOS 6.10 VMware test system. _(My apologies for not having tested my original patch there before asking you to test it! I guess I got busy and forgot!)_

If it builds and runs okay for you this time _(it should!)_ then I'll go ahead and commit it and then close this issue.

_**But please respond to Peter's concerns too!** &nbsp; (please?)_

He specifically mentioned his concern about your having dropped out of the conversation / discussion after my having responded to him.

I'm not sure what (if anything) we need/should/can do to make him happy, but as I'm sure you know (and agree with too), he's such an important long time contributor to the Hercules project that IMO we should at least _try_ to meet any _reasonable_ demand of his to in order to make him happy/comfortable with what we're doing with his code/contribution to Hercules (which he doesn't seem to be right now which is causing me considerable concern).

As for me, I believe I've done all that I can.

Maybe you _(or someone else?)_ can do more?

Thanks.

**Comment by ivan-w** [comment]
At this point I'm thinking that I'm going to start from scratch.
In the process I'm also going to do a "dummy" device driver that can be used as a skeleton for future device handler developers as well as an I/O exerciser.

**Comment by Fish-Git** [comment]
Sounds good to me!

**Comment by Juergen-Git** [comment]
Hi Ivan

Given that I currently have put the TK4- release process on hold waiting for #289 to be fixed, I intend to use the wait time to integrate Bob's NJE38 product into the system. Thus the shy question: Where do you stand with the TCPNJE port?

Cheers
Jürgen

**Comment by agadsby** [comment]
Hi Jurgen and team 

I’ve put the fix in for the select CPU burn  issue. I believe that is sufficient to fix the reported issue. I’m working through the logger thread stuff now. Oddly keep the logger thread alive for longer during shutdown removes a lot of complexity. It’s going to be a few days before I’m happy with proposing some further changes. 

Andy

**Comment by Fish-Git** [comment]
Andrew Gadsby (@agadsby) wrote:

> Hi Jurgen and team.
>
> I’ve put the fix in for the select CPU burn issue. I believe that is sufficient to fix the reported issue.

The fix is in _your personal fork_ of SDL Hyperion, yes. I can see that from the helpful GitHub mention in issue #289. But your fix is still _not_ in the official SDL Hyperion repository yet. It's only in _your_ repository (fork), but not in _our_ repository. &nbsp; `:)`

I've never done it myself so I'm not one to speak, but I believe there's a way to submit/create a "pull request" requesting that _we_ (the SDL Hyperion repository) please do a "pull" _from **your**_ repository (fork) of the commit that you made that fixed the problem. Then we could Merge your pull request to get your fix into our repository and you'd get credit for it.

**_OR...._** &nbsp; The much _faster_ and _easier_ way (easier for both you and me) would be, since the fix is so simple, for me to simply directly commit your fix to our repository myself _(but with you as the "Author" of the fix of course, so that you get proper credit for it)_.

So I think that's what I'm going to do. &nbsp; `;-)`

Give me a few minutes and I'll commit your fix and close this issue.

**Comment by agadsby** [comment]
Thank you that’s a great plan. Once I have the logger changes I’ll try and do it properly  on a new issue. 

Andy

**Comment by Fish-Git** [comment]
> Give me a few minutes and I'll commit your fix and close this issue.

_(Oops!)_ Wrong issue!

Previous comments moved to Issue #289 where they belong.

Sorry about that folks!

**Comment by Fish-Git** [comment]
> ... I currently have put the TK4- release process on hold waiting for #289 to be fixed ...

Jürgen:

Now that #289 has been fixed (closed) you may now release your new TK4- at your earliest convenience. Thanks.

**Comment by Juergen-Git** [comment]
Hi Fish

It's not _that_ easy!

- Firstly, the fix for #289 solves one specific consequence of our message pipe implementation only, by "suppressing" the symptom. However, as I stated a few times, I'm convinced there are other issues that also have their root causes in message handling (comm3705 and commadpt instabilities, console polling anomalies, etc.). So, I'd prefer to see Andy's complete solution before publishing TK4- in a hurry and having to come up with new binaries shortly after -- as you know, I'm supporting many platforms and it's quite a task for me to build and test the binaries for all of them.

- Secondly, TCPNJE is on my watchlist for quite some time now. I had this on high priority already in last autumn, but switched priorities to wait for a ready to use solution to get NJE functionality into TK4-. The funetnje approach I published in autumn was meant to be temporary from the beginning. Now that Bob's NJE38 is ready for prime time, the unclear situation around the TCPNJE driver is the only obstacle hindering me to add NJE38 to TK4- Update 09. Thus my question above concerning its status.

So, I will not hurry at the moment and go after the MVS side of implementing NJE38 in TK4- for now...

Cheers
Jürgen

**Comment by Fish-Git** [comment]
Jürgen wrote:
> Fish wrote:
> > Jürgen wrote:
> 
> > > ... I currently have put the TK4- release process on hold waiting for #289 to be fixed ...
> 
> > Now that #289 has been fixed (closed) you may now release your new TK4- at your earliest convenience. Thanks.
> 
> Hi Fish
> 
> It's not _that_ easy!
> 
> * Firstly, the fix for #289 solves one specific consequence of our message pipe implementation only, by "suppressing" the symptom. However, as I stated a few times, I'm convinced there are other issues that also have their root causes in message handling (comm3705 and commadpt instabilities, console polling anomalies, etc.). So, I'd prefer to see Andy's complete solution before publishing TK4- in a hurry and having to come up with new binaries shortly after ...

Understood! That's why I said "at your earliest convenience". That means: "whenever you feel it's appropriate to do so". There's no rush! I was just letting you know that one of your stated road blocks had been removed and so you were now free to proceed.

But if you want to wait until Andy has a more complete/better fix to the message pipe problem before releasing your code, that's fine! As I said, no rush. It's your call. I'm fine with whatever you decide.

**Comment by agadsby** [comment]
Hi Fish and Jurgen

I think I now understand what is causing the other issues, the pipe is blocking under load which will cause mayhem for the calling thread. I have a fix in my head for this which I will test out and then upload to a new issue early next week. I will also include my general logger cleanup as well. 

Andy

**Comment by Fish-Git** [comment]
Andrew Gadsby (@agadsby) wrote:

> Hi Fish and Jurgen I think I now understand what is causing the other issues, the pipe is blocking under load ...

Wrong thread. &nbsp; (i.e. Wrong GitHub Issue.)

THIS issue is #280, and has to do with Porting TCPNJE support from spinhawk to SDL Hyperion.

The issue you should be posting your comments to is issue #289, the Panel issue under Linux.

Also, it would be greatly appreciated if you would please NOT reply to GitHub Issues via email. Rather, it would be greatly appreciated if you would please instead logon to GitHub and reply directly from GitHub itself: &nbsp; **---&gt; https://github.com/SDL-Hercules-390/hyperion/issues**.

(Don't worry that issue #289 is closed. That's okay. You can still post comments to closed issues)

**SUMMARY:**
1. Please do NOT reply to github issues via email.
2. Please DO reply to github issues directly via github.
3. You can still comment to closed issues.

_Thanks!_

**Comment by Fish-Git** [comment]
Fish wrote:
> Jürgen wrote:
> 
> > Hi Fish
> > It's not _that_ easy!
> > 
> > * Firstly, the fix for #289 solves one specific consequence of our message pipe implementation only, by "suppressing" the symptom. However, as I stated a few times, I'm convinced there are other issues that also have their root causes in message handling (comm3705 and commadpt instabilities, console polling anomalies, etc.). So, I'd prefer to see Andy's complete solution before publishing TK4- in a hurry and having to come up with new binaries shortly after ...
> 
> Understood! That's why I said "at your earliest convenience". That means: "whenever you feel it's appropriate to do so". There's no rush! I was just letting you know that one of your stated road blocks had been removed and so you were now free to proceed.
> 
> But if you want to wait until Andy has a more complete/better fix to the message pipe problem before releasing your code, that's fine! As I said, no rush. It's your call. I'm fine with whatever you decide.

Hi Jürgen! (@Juergen-Git)

What's the current status of this issue? (Just asking!)

AFAIK, the fix that was committed to close #289 _**is_** Andy's "complete solution". Yes?

Thanks.

**Comment by Juergen-Git** [comment]
Hi Fish

from my point of view, this one can be closed, [208e766](url) is stable for many weeks now.

Cheers
Jürgen

**Comment by Fish-Git** [comment]
Closing.
